### DISCUSSION TITLE
Writing a copy of a heavystruct

### DISCUSSION THREAD

**Discussion by solidkon** [mupif/mupif#70] [Q&A] [answered]
Hello!

I will describe a scenario and ask the question later. Let's suppose that we run a MD simulation of a collection of molecules that are stored inside a hdf5 container. After the simulation is finished we have a new state of the container (the one so called GrainState) and we want to store the data inside a new hdf5 container and keep the old one for possible future usage, i.e. we do not want to update the container. What is the recommended methodology to accomplish this? I have 2 possibilities in mind:

1) Get a deepcopy of the initial container, update the desired attributes and the dump it to a new hdf5 file
2) Initialize a new grain (this requires the schema to be loaded that is stored inside the initial container) and then store the desired data inside.

I see that there are several bound methods for the mupif.heavystruct objects. I'm just not quite sure which/how to use the desired one properly.

**Comment by eudoxos** [comment]
Hi Kostis,

we discussed this internally a few times. Updating the input container is definitely not something we want to do, that would destroy the traceability of the simulation (the ability to associate any simulation run with its input values). When you open the hdf5 container and modify the data, they are written (more or less) immediately into the hdf5 file.

For cases when you want to have slightly modified input data as output, we recently overloaded the `HeavyData.deepcopy` method so that it will also copy the backing hdf5 container (when called locally) or turn proxy to the remote hdf5 container into a local one (the copy of the container runs over the network). So call deepcopy first, then do whatever you want with the container.

If you are creating a new grain state, create a new container for output. For schemata JSON, simply re-use the input's `schemasJson` attribute when creating the new `HeavyStruct`.

**Reply by solidkon** [reply]
Hello eudoxos!

Thanks a lot for your answer. I would like to ask something else also, relevant to the strategy I have chosen for the MD API. I will probably go with a new grain initialization. Is there a way to retrieve the schema stored inside a container? This would make things much easier, since I would not need to have the json file every time I'm planning to use a grain hdf5.

**Comment by eudoxos** [comment]
> Thanks a lot for your answer. I would like to ask something else also, relevant to the strategy I have chosen for the MD API. I will probably go with a new grain initialization. Is there a way to retrieve the schema stored inside a container? This would make things much easier, since I would not need to have the json file every time I'm planning to use a grain hdf5.

What you write makes sense. The JSON is stored in the container, plus loaded into `HeavyStruct.schemasJson` — but we will have to add a method to make it accessible remotely, e.g. `HeavyStruct.getSchemasJson()`. Sounds good?

**Reply by solidkon** [reply]
Yes it sounds good. For the moment I pass the string corresponding to schema as a model attribute. I didn't get into more trouble to make it a pyrofile since it is a temporary solution.

**Reply by eudoxos** [reply]
Good, `HeavyStruct.getSchemasJson` and `HeavyStruct.getSchemaName` was added in master; it will take a few days before it percolates in the `Musicode` branch.

**Comment by solidkon** [comment]
Hello eudoxos! I am testing the new bound methods for Heavystructs, i.e. the `HeavyStruct.getSchemasJson()` and `HeavyStruct.getSchemaName()`. However, the return type of these two is of `NoneType`. I thought that this information was stored inside the container independently of how a Heavystruct object was instantiated through an existing .h5 container. Am I wrong here? What I mean is that when I am instantiating an existing .h5 container, I do not know the schema used for creating this container in the first place. I assume that the schema does not change and is strictly defined in the initial construction. The instantiation is this:

`ga_hs = mp.HeavyStruct(h5path='assembly_in.h5', mode='readonly', id=mp.DataID.ID_GrainState)`

As you can see the heavystruct object is created but neither the `SchemaName` nor the` SchemasJson` arguments are passed in the construction. Is this why am I getting the NoneType as return type for the bound methods?

The general idea here is that the container is initialized once at the very beginning of a workflow. This will allow a universal schema definition only once. From this point every new container will be described by the same schema. And the aforementioned bound methods will serve this purpose. Am I missing something here?

**Comment by eudoxos** [comment, marked as answer]
Hi Kostis, sorry for this —  the `getSchemasJson()` will only work on a newly-constructed `HeavyStruct`. That is a problem which I overlooked, will fix that soon.

When loading a `HeavyStruct`, with the bug fixed, you will sill need to open the data (via `openData` or via context manager) first so that the schemata stored in the container will be accessible.

**Reply by solidkon** [reply]
Thanks a lot eudoxos! This is exactly what will be needed in all cases! Just send me when the branch has been updated in order to adjust the code appropriately.

**Reply by eudoxos** [reply]
Hi fcc6063 should do the trick, please try.

**Reply by solidkon** [reply]
Yes it works as expected!!! I can retrieve the schema with the context manager. Thank you so much for your immediate response

**Reply by solidkon** [reply]
Hi eudoxos, do you know when these changes will be applied to mupif repository? I was able to do it two days ago on my home pc but now the `git pull` command states that there are divergent branches.

**Reply by eudoxos** [reply]
Hi, sorry for overlooking the reply. Could you solve it? We had some mess merging changes to Musicode but it should be alright now — do you still have the issue?

**Reply by solidkon** [reply]
Unfortunately the issue still persists when I'm attempting a `git pull` inside mupif repository.

**Reply by solidkon** [reply]
Eudoxos sorry I just reinstalled mupif and everything works properly!
